taskfiles/utils.yml

```
@@ -141,3 +148,158 @@ tasks:
+    internal: true
+    vars:
+      CMAKE_ARGS: >-
+        {{default "" .CMAKE_ARGS}}
+    requires:
+      vars: ["BUILD_DIR", "SOURCE_DIR"]
+    cmds:
+      - >-
+        cmake
+        -S "{{.SOURCE_DIR}}"
+        -B "{{.BUILD_DIR}}"
+        {{.CMAKE_ARGS}}
+      - >-
+        cmake
+        --build "{{.BUILD_DIR}}"
+        --parallel
+
+  # Unconditionally runs the CMake install step for the given build directory.
+  #
+  # @param {string} BUILD_DIR CMake build directory.
+  # @param {string} INSTALL_PREFIX Path prefix of where the project should be installed.
+  cmake-install:
+    label: "{{.TASK}}-{{.BUILD_DIR}}-{{.INSTALL_PREFIX}}"
+    internal: true
+    requires:
+      vars: ["BUILD_DIR", "INSTALL_PREFIX"]
+    cmds:
+      - >-
+        cmake
+        --install "{{.BUILD_DIR}}"
+        --prefix "{{.INSTALL_PREFIX}}"
+
+  # ===
+  # REMOTE UTILS
+  # ===
+
+  # Runs curl to download a file from the given URL.
+  #
+  # @param {string} URL
+  # @param {string} FILE_SHA256 Content hash to verify downloaded file against.
+  # @param {string={{(base .URL)}}} [OUTPUT_FILE] Path where the file should be stored.
+  curl:
+    label: "{{.TASK}}-{{.OUTPUT_FILE}}"
+    internal: true
+    vars:
+      OUTPUT_FILE: "{{default (base .URL) .OUTPUT_FILE}}"
+    requires:
+      vars: ["URL", "FILE_SHA256"]
+    generates: ["{{.OUTPUT_FILE}}"]
+    status:
+      - >-
+        diff
+        <(echo "{{.FILE_SHA256}}")
+        <(openssl dgst -sha256 "{{.OUTPUT_FILE}}"
+        | awk '{print $2}')
+    cmds:
+      - |-
+        mkdir -p "{{dir .OUTPUT_FILE}}"
+        max_attempts=3
+        attempt=1
+        while [ $attempt -le $max_attempts ]; do
+          if curl \
+              --fail \
+              --location \
+              --show-error \
+              --connect-timeout 10 \
+              --max-time 300 \
+              "{{.URL}}" \
+              --output "{{.OUTPUT_FILE}}";
+          then
+            break
+          fi
+          attempt=$((attempt + 1))
+          sleep 5
+        done
+        if [ $attempt -gt $max_attempts ]; then
+          echo "Failed to download after $max_attempts attempts."
+          exit 1
+        fi
+
+  # Runs curl to download the tarball file from the given URL and extracts its contents.
+  #
+  # @param {string} OUTPUT_DIR Directory in which to extract the tarball file.
+  # @param {string} URL
+  # @param {string} FILE_SHA256 Content hash to verify downloaded tar file against.
+  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of
+  # downloaded tar file.
+  # @param {string[]=[]} [EXCLUDE_PATTERNS] Path wildcard patterns that should not be extracted.
+  # @param {string[]=[]} [DATA_PATTERNS] Path wildcard patterns to extract.
+  # @param {int=1} [NUM_COMPONENTS_TO_STRIP] Number of leading path components to strip from the
+  # extracted files.
+  # @param {string={{.OUTPUT_DIR}}.tar.gz} [TAR_FILE] Path where the tarball file should be stored.
```

```suggestion
  # Runs curl to download the tar file from the given URL and extracts its contents.
  #
  # @param {string} OUTPUT_DIR Directory in which to extract the tar file.
  # @param {string} URL
  # @param {string} FILE_SHA256 Content hash to verify the downloaded tar file against.
  # @param {string={{.OUTPUT_DIR}}.md5} [CHECKSUM_FILE] File path to store the checksum of
  # downloaded tar file.
  # @param {string[]=[]} [EXCLUDE_PATTERNS] Path wildcard patterns that should not be extracted.
  # @param {string[]=[]} [DATA_PATTERNS] Path wildcard patterns to extract.
  # @param {int=1} [NUM_COMPONENTS_TO_STRIP] Number of leading path components to strip from the
  # extracted files.
  # @param {string={{.OUTPUT_DIR}}.tar.gz} [TAR_FILE] Path where the tar file should be stored.
```

Upon re-reading, I realized `tar file` is better.